File: main/src/main/java/org/anibyl/slounik/network/Article.java
package org.anibyl.slounik.network;

import android.text.Spanned;

/**
 * Article.
 * <p/>
 * Created by Usievaład Čorny on 26.02.2015 14:06.
 */
public abstract class Article {
    private final DictionarySiteCommunicator communicator;
    private String title;
    private Spanned description;
    private String dictionary;
    private String linkToFullDescription;
    private Spanned fullDescription;

    public Article(DictionarySiteCommunicator communicator) {
        this.communicator = communicator;
    }

    public String getTitle() {
        return title;
    }

    public Spanned getDescription() {
        return description;
    }

    public String getDictionary() {
        return dictionary;
    }

    public String getLinkToFullDescription() {
        return linkToFullDescription;
    }

    public Spanned getFullDescription() {
        return fullDescription;
    }

    public DictionarySiteCommunicator getCommunicator() {
        return communicator;
    }

    abstract Article fill();

    Article setTitle(String title) {
        this.title = title;
        return this;
    }

    Article setDescription(Spanned description) {
        this.description = description;
        return this;
    }

    Article setDictionary(String dictionary) {
        this.dictionary = dictionary;
        return this;
    }

    Article setLinkToFullDescription(String linkToFullDescription) {
        this.linkToFullDescription = linkToFullDescription;
        return this;
    }

    void setFullDescription(Spanned fullDescription) {
        this.fullDescription = fullDescription;
    }
}

File: main/src/main/java/org/anibyl/slounik/network/Article.kt
package org.anibyl.slounik.network

import android.text.Spanned
import org.anibyl.slounik.core.fromHtml
import java.io.Serializable

/**
 * Article.
 *
 * @author Usievaład Kimajeŭ
 * @created 26.02.2015
 */
class Article(@Transient val communicator: DictionarySiteCommunicator) : Serializable {
	var title: String? = null
		internal set
	var description: String? = null
		internal set
	var dictionary: String? = null
		internal set
	var linkToFullDescription: String? = null
		internal set
	var fullDescription: String? = null
		internal set

	val spannedDescription: Spanned?
		get() = description?.fromHtml()
	val spannedFullDescription: Spanned?
		get() = fullDescription?.fromHtml()
}

fun Article.loadArticleDescription(callback: (ArticlesInfo) -> Unit) {
	this.communicator.loadArticleDescription(this, object : ArticlesCallback {
		override fun invoke(info: ArticlesInfo) = callback(info)
	})
}
